### PR TITLE
Add new BlockPolarForm formulation

### PR DESCRIPTION
**Pull request by frapac**
This new structure is a drop-in replacement for the former `BlockNetworkStack`: the block structure is now given directly in the formulation. 

As a consequence, this PR: 
- remove the `BlockNetworkStack` structure (which was almost a duplicate of `NetworkStack`)
- remove the `blockmul!` function to use `LinearAlgebra.mul!`

### CODE REVIEW

File: docs/src/tutorials/batch_evaluation.md [hunk 1]
```diff
@@ -42,24 +42,38 @@ pd = stack.params[1:nbus]
...
+formulation for $$N$$ different set of parameters (=scenarios) $$p_1, \cdots, p_N$$.
+ExaPF allows to streamline the polar formulation with a [`BlockPolarForm`](@ref)
+structure:
 ```@example batch_pf
 nscen = 10;
+blk_polar = ExaPF.BlockPolarForm(polar, nscen)
+
+```
+Then, ExaPF can also instantiate a [`NetworkStack`](@ref)
+object, with the memory required to store the variables on
```
Comment (michel2323) [inline docs/src/tutorials/batch_evaluation.md:57]:
```suggestion
object, with the memory required to store the variables of
```

File: docs/src/tutorials/batch_evaluation.md [hunk 2]
```diff
@@ -42,24 +42,38 @@ pd = stack.params[1:nbus]
...
 
 ```
-In short, a [`NetworkStack`](@ref) stores one set of loads $$p_0$$.
-On its side, the [`BlockNetworkStack`](@ref) extends the `NetworkStack` structure
-to store $$N$$ different set of parameters (=scenarios) $$p_1, \cdots, p_N$$.
-As an illustration, the syntax to instantiate a [`BlockNetworkStack`](@ref)
-with $$N=10$$ different scenarios is:
+By default, a [`NetworkStack`](@ref) stores one set of loads $$p_0$$.
+
+Suppose now we want to evaluate the model associated to the polar
```
Comment (michel2323) [inline docs/src/tutorials/batch_evaluation.md:47]:
```suggestion
Suppose now we want to evaluate the model associated with the polar
```

File: docs/src/tutorials/batch_evaluation.md [hunk 3]
```diff
@@ -72,26 +86,17 @@ reshape(blk_stack.pload, nbus, nscen)
 
 ## Evaluate expressions in block
 
-A [`BlockNetworkStack`](@ref) has the same behavior as a
-[`NetworkStack`](@ref) structure, and any
-[`AutoDiff.AbstractExpression`](@ref) can take a [`BlockNetworkStack`](@ref)
-as input. ExaPF can takes advantage of the block structure
-when using a [`BlockNetworkStack`](@ref).
+ExaPF can takes advantage of the block structure when using a [`BlockPolarForm`](@ref).
```
Comment (michel2323) [inline docs/src/tutorials/batch_evaluation.md:89]:
```suggestion
ExaPF takes advantage of the block structure when using a [`BlockPolarForm`](@ref).
```

File: docs/src/tutorials/batch_evaluation.md [hunk 4]
```diff
@@ -72,26 +86,17 @@ reshape(blk_stack.pload, nbus, nscen)
...
 
 As an example, suppose we want to evaluate the power flow
 balances in block with a [`PowerFlowBalance`](@ref) expression:
```
Comment (michel2323) [inline docs/src/tutorials/batch_evaluation.md:92]:
```suggestion
balances in block form with a [`PowerFlowBalance`](@ref) expression:
```

File: src/Polar/polar.jl [hunk 5]
```diff
@@ -24,7 +33,7 @@ giving a mathematical formulation with:
 
 ```
 """
-struct PolarForm{T, IT, VT, MT} <: AbstractFormulation where {T, IT, VT, MT}
+struct PolarForm{T, IT, VT, MT} <: AbstractPolarFormulation{T, IT, VT, MT}
```
Comment (michel2323) [inline src/Polar/polar.jl:36]: What difference does this change make?

File: src/Polar/polar.jl [hunk 6]
```diff
@@ -36,192 +45,32 @@ end
...
-"""
-function load_polar(case, device=CPU(); dir=PS.EXADATA)
-    return PolarForm(PS.load_case(case, dir), device)
-end
-
-# Getters (bridge to PowerNetwork)
-get(polar::PolarForm, attr::PS.AbstractNetworkAttribute) = get(polar.network, attr)
-
-number(polar::PolarForm, v::AbstractVariable) = length(mapping(polar, v))
+name(polar::PolarForm) = "Polar formulation"
```
Comment (michel2323) [inline src/Polar/polar.jl:48]: Cute :-) .

File: src/Polar/second_order.jl [hunk 7]
```diff
@@ -133,10 +133,10 @@ function FullHessian(polar::PolarForm{T, VI, VT, MT}, func::AutoDiff.AbstractExp
     H = H_host |> SMT
 
     # Structures
-    stack = NetworkStack(nbus, ngen, nlines, VT, VD)
+    stack = NetworkStack(nbus, ngen, nlines, 1, VT, VD)
```
Comment (michel2323) [inline src/Polar/second_order.jl:136]: Can't we have a default constructor here for `1`?

File: src/Polar/utils.jl [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+5-element Vector{Int64}:
+  1
+  2
+  3
+ 20
+ 21
+
+```
+
+"""
```
Comment (michel2323) [inline src/Polar/utils.jl:50]: `mapping` is in `utils.jl`? (maybe it is...)

**Review by michel2323 — APPROVED** [review]
Looks good to me!

File: src/Polar/polar.jl (hunk 5 above)
Comment (frapac) [reply src/Polar/polar.jl:36]: We have a new abstract type `AbstractPolarFormulation` that encapsulates both `PolarForm` and `BlockPolarForm`

**Pull request merged**